src/ConformalTracking.cc

```diff
@@ -1667,6 +1667,341 @@ void ConformalTracking::extendHighPT(UniqueKDTracks& conformalTracks, SharedKDCl
+      double theta = kdhit->getTheta();
+      nearestNeighbours->allNeighboursInTheta(
+          theta, m_thetaRange * 4, results, [&kdhit, vertexToTracker](SKDCluster const& nhit) {
+            //if same subdet, take only the hits within two layers
+            if (nhit->getSubdetector() == kdhit->getSubdetector()) {
+              if ((vertexToTracker && nhit->getLayer() > (kdhit->getLayer() + 2)) ||
+                  (!vertexToTracker && nhit->getLayer() < (kdhit->getLayer() - 2))) {
+                return true;
+              }
+            }
+
+            if ((vertexToTracker && nhit->getR() > kdhit->getR()) || (!vertexToTracker && nhit->getR() < kdhit->getR()))
+              return true;
+            return false;
+          });
+      //nearestNeighbours->allNeighboursInRadius(kdhit, parameters._maxDistance, results);
+      streamlog_out(DEBUG9) << "- Found " << results.size() << " neighbours. " << std::endl;
+      if (streamlog::out.write<DEBUG9>()) {
+        for (unsigned int neighbour = 0; neighbour < results.size(); neighbour++) {
+          streamlog_out(DEBUG9) << "-- Neighbour from allNeighboursInTheta " << neighbour << ": [x,y] = ["
+                                << results.at(neighbour)->getX() << ", " << results.at(neighbour)->getY()
+                                << "]; r = " << results.at(neighbour)->getR() << "; [subdet,layer] = ["
+                                << results.at(neighbour)->getSubdetector() << ", " << results.at(neighbour)->getLayer()
+                                << "]" << std::endl;
+        }
+      }
+
+      if (results.size() == 0) {
+        loop = false;
+        continue;
+      }
+
+      std::sort(results.begin(), results.end(), (vertexToTracker ? sort_by_layer : sort_by_lower_layer));
+      // Get the final values of subdet and layer to stop the loop at the (vertexToTracker? outermost : innermost) layer with neighbours
+      final_subdet = results.at(results.size() - 1)->getSubdetector();
+      final_layer  = results.at(results.size() - 1)->getLayer();
+      streamlog_out(DEBUG9) << "- Final layer with neighbours: [subdet,layer] = [" << final_subdet << ", " << final_layer
+                            << "]" << std::endl;
+
+      if (streamlog::out.write<DEBUG9>()) {
+        for (unsigned int neighbour = 0; neighbour < results.size(); neighbour++) {
+          streamlog_out(DEBUG9) << "-- Neighbour " << neighbour << ": [x,y] = [" << results.at(neighbour)->getX() << ", "
+                                << results.at(neighbour)->getY() << "]; [subdet,layer] = ["
+                                << results.at(neighbour)->getSubdetector() << ", " << results.at(neighbour)->getLayer()
+                                << "]" << std::endl;
+        }
+      }
+
+      // If no hit was found on a layer, we use expectedHit as current position
+      // However, the nearest neighbour search is always performed from the last real hit (kdhit)
+      if (!skipLayer)
```

Comment:
The definition of `hitOnCurrentLayer` can be moved here (and made a reference, I think)
`SKDCluster& hitOnCurrentLayer = skipLayer ? expectedHit : kdhit;`

Reply:
nice